File: Python/Deque.py
class Deque:

  def __init__(self, n):
    self.deque = [0]*n
    self.front = n+1
    self.rear = 0
    self.maxSize = n
    self.currentSize = 0

  
  def isFull(self):
    return self.currentSize == self.maxSize
  
  def isEmpty(self):
    return self.currentSize  == 0

  
  def addFront(self, data):
    if self.isFull():
      print("OPERATION FAILED, DEQUE FULL")
      return
    
    self.front = (self.front - 1)%self.maxSize
    self.deque[self.front] = data
    self.currentSize +=1
  
  def addRear(self, data):
    if self.isFull():
      print("OPERATION FAILED, DEQUE FULL")
      return
    
    self.rear = (self.rear + 1)%self.maxSize
    self.deque[self.rear] = data
    self.currentSize+=1
  
  def removeRear(self):
    if self.isEmpty():
      print("OPERATION FAILED, DEQUE EMPTY")
      return
    data = self.deque[self.rear]
    self.deque[self.rear] = 0
    self.rear = (self.rear - 1) % self.maxSize
    self.currentSize -=1
    return data

  def removeFront(self):
    if self.isEmpty():
      print("OPERATION FAILED, DEQUE EMPTY")
      return

    data = self.deque[self.front]
    self.deque[self.front] = 0
    self.front = (self.front + 1) % self.maxSize
    self.currentSize -=1
    return data

  def peekFront(self):
    if self.isEmpty():
      print("OPERATION FAILED, DEQUE EMPTY")
      return
    
    return self.deque[self.front]

  def peekRear(self):
    if self.isEmpty():
      print("OPERATION FAILED, DEQUE EMPTY")
      return

    return self.deque[self.rear]





def main():
  dq = Deque(3)
  print('Insert Element at front')
  print("Insert element at rear end  : 5 ")
  dq.addRear(5)
  print("Insert element at rear end : 10 ")
  dq.addRear(10)
  print(f"Get rear element : {dq.peekRear()}")
  dq.removeRear()
  print(f"After delete rear element new rear become : {dq.peekRear()}")
  print("Inserting element at front end")
  dq.addFront(15)
  print(f"Get front element: {dq.peekFront()}")
  dq.removeFront()
  print(f"After delete front element new front become : {dq.peekFront()}")




  
main()

File: Java/Deque.java
class Deque {
	int[] deque;
	int rear, front, currentSize, maxSize;

	Deque(int n) {
		deque = new int[n];
		rear = 0;
		front = n + 1;
		maxSize = n;
		currentSize = 0;
	}

	boolean isFull() {
		return (currentSize == maxSize);
	}

	boolean isEmpty() {
		return (currentSize == 0);
	}

	void addFront(int data) {
		if (isFull()) {
			System.out.println("Operation Failed, DEQUE FULL");
			return;
		}
		front = (front - 1) % maxSize;
		front = (front == -1) ? maxSize - 1 : front;
		deque[front] = data;
		currentSize++;
	}

	void addRear(int data) {
		if (isFull()) {
			System.out.println("Operation Failed, DEQUE FULL");
			return;
		}
		rear = (rear + 1) % maxSize;
		deque[rear] = data;
		currentSize++;
	}

	int removeFront() {
		if (isEmpty()) {
			System.out.println("Operation Failed, DEQUE EMPTY");
			return (-maxSize - 1);
		}
		int data = deque[front];
		deque[front] = 0;
		front = (front + 1) % maxSize;
		currentSize--;
		return data;

	}

	int removeRear() {
		if (isEmpty()) {
			System.out.println("Operation Failed, DEQUE EMPTY");
			return (-maxSize - 1);
		}
		int data = deque[rear];
		deque[rear] = 0;
		rear = (rear - 1) % maxSize;
		rear = (rear == -1) ? maxSize - 1 : rear;

		currentSize--;
		return data;

	}

	int peekFront() {
		if (isEmpty()) {
			System.out.println("Operation Failed, DEQUEUE EMPTY");
			return (-maxSize - 1);
		}
		return deque[front];
	}

	int peekRear() {
		if (isEmpty()) {
			System.out.println("Operation Failed, DEQUEUE EMPTY");
			return (-maxSize - 1);
		}
		return deque[rear];
	}

	void print() {
		for (int i = 0; i < maxSize; i++) {
			System.out.print(deque[i] + " ");
		}
		System.out.println("\nFront: " + front + "\tRear: " + rear);
	}

	public static void main(String[] args) {
		Deque dq = new Deque(3);
		System.out.println("Insert element at rear end  : 5 ");
		dq.addRear(5);
		System.out.println("Insert element at rear end : 10 ");
		dq.addRear(10);
		System.out.println("Get rear element : " + dq.peekRear());
		dq.removeRear();
		System.out.println("After delete rear element new rear become : " + dq.peekRear());
		System.out.println("inserting element at front end");
		dq.addFront(15);
		System.out.println("get front element: " + dq.peekFront());
		dq.removeFront();
		System.out.println("After delete front element new front become : " + +dq.peekFront());
	}

}
